packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,415 @@
+          .filter(([namespaceId]) => resolvedAccounts[namespaceId]?.accounts)
+          .map(([namespaceId, namespace]) => [
+            namespaceId,
+            {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+              accounts: resolvedAccounts[namespaceId]!.accounts,
+              chains: namespace.chains,
+              events: namespace.events,
+              methods: namespace.methods,
+            },
+          ]),
+      );
+
+    const session: SessionData = {
+      origin,
+      requestedNamespaces: connection.requiredNamespaces,
+      providedNamespaces,
+      handlingSnaps: fromEntries(
+        Object.entries(resolvedAccounts)
+          .filter(([, data]) => data !== null)
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          .map(([namespaceId, data]) => [namespaceId, data!.snapId]),
+      ),
+    };
+
+    await Promise.all(
+      Object.values(session.handlingSnaps).map((snapId) =>
+        this.messagingSystem.call('SnapController:onSessionOpen', snapId),
+      ),
+    );
+
+    this.update((state) => {
+      state.sessions[origin] = session;
+    });
+
+    return { namespaces: providedNamespaces };
+  }
+
+  async onRequest(
+    origin: string,
+    data: { chainId: ChainId; request: RequestArguments },
+  ): Promise<unknown> {
+    const session = this.getSession(origin);
+    assert(session, `Session for "${origin}" doesn't exist`);
+    const { namespace } = parseChainId(data.chainId);
+    assert(
+      session.providedNamespaces[namespace]?.chains.includes(data.chainId),
+      `Session for "${origin}" not connected to "${data.chainId}" chain`,
+    );
+    const snapId = session.handlingSnaps[namespace];
+    assert(snapId !== undefined);
+    const permissionName = getSnapPermissionName(snapId);
+
+    // Check if origin has permission to communicate with this Snap.
+    const hasPermission = await this.messagingSystem.call(
+      'PermissionController:hasPermission',
+      origin,
+      permissionName,
+    );
+
+    if (!hasPermission) {
+      // TODO: Get permission for origin connecting to snap, or get user approval
+      // In the future this is where we should prompt for this permission.
+      // In this iteration, we will grant this permission in onConnect.
+      throw new Error(
+        `${origin} does not have permission to communicate with ${snapId}`,
+      );
+    }
+
+    return this.snapRequest({
+      snapId,
+      origin,
+      method: 'handleRequest',
+      args: data,
+    });
+  }
+
+  private async snapRequest({
+    snapId,
+    origin,
+    method,
+    args,
+  }: {
+    snapId: string;
+    origin: string;
+    method: string;
+    args?: unknown;
+  }) {
+    return this.messagingSystem.call('SnapController:handleRequest', {
+      snapId,
+      origin,
+      handler: HandlerType.SnapKeyring,
+      request: { method, params: args ? [args] : [] },
```

Comment:
`args ? [args] : []` should probably be done in SnapController

Reply:
Not sure I agree 🤔 

This is specifically a keyring thing and doing it in the SnapController will be more complicated than this.